### DISCUSSION TITLE
[Feature Request] Infinite Scroll

### DISCUSSION THREAD

**Discussion by pvanthony** [GetPublii/Publii#1421] [Ideas]
Could we have infinite Scroll for publii websites?

Here is a how to for wordpress.
https://www.wpbeginner.com/plugins/how-to-add-infinite-scroll-to-your-wordpress-blog/

Here is a news website that has infinite scroll. I hope infinite scroll is the correct name. 
https://www.channelnewsasia.com/world/pope-says-ukraine-should-have-courage-white-flag-negotiations-4183841

**Comment by candidexmedia** [comment]
Some themes have it already:

Sources: https://demo.getpublii.eu/themes/sources/2/

Persona (premium): https://marketplace.getpublii.com/themes/persona/

Luckily, you can implement Infinite scroll yourself by adding the JS library to your site files: https://infinite-scroll.com/


---

Similar topics from the old forum:
- https://forum.getpublii.com/topic/mercury-theme-filter-filter-all-posts/
- https://forum.getpublii.com/topic/does-publii-have-infinite-loading-feature

**Reply by pvanthony** [reply]
Thank you very much for the reply and the links. 
Very useful!

So I just have to choose a theme that has the infinite scroll feature.
I have to look for a theme that will do infinite scroll for all the blog posts.

Thank you very much for the sharing the information. It is very useful because it points me to the right path of checking the themes.

**Reply by candidexmedia** [reply]
No problem!

You can either select a theme that already has it, *or* manually add infinite scroll to the theme of your choice.

**Reply by pvanthony** [reply]
That's the problem. I do not know how to add it to the theme.
Currently I am using Editorial 2 version 1.4.0.0
https://marketplace.getpublii.com/themes/editorial-2/

**Reply by candidexmedia** [reply]
I tried testing it on my end, but got stuck and couldn't get it to work...

In the meantime, changing your "Posts per page" Theme setting to "-1" can also work to display all your posts at once. That's what I've done in another blog, since I didn't want to move to another theme just for infinite scroll.

![image](https://github.com/GetPublii/Publii/assets/95003834/d9710f6f-0059-4b30-95fd-e14e2dd11eb0)


I'll share what I tried so far for infinite scroll in Editorial 2 anyway, in case someone else has better luck than me: 

# Add the Infinite Scroll .js 

Easiest way to do this is to go to Tools & Plugins > Custom HTML > Footer and paste the following:

```js
<!--- INFINITE SCROLL JAVASCRIPT --->
<script src="https://unpkg.com/infinite-scroll@4/dist/infinite-scroll.pkgd.min.js"></script>
```

# Initialize the JS

Infinite Scroll works on a container element with item elements. In your Editorial 2's case, it looks something like this:

```html
<div class="posts">
  <article>...</article>
  <article>...</article>
  <article>...</article>
  <article>...</article>
  ...
</div>
```

There are two ways to activate infinite scroll in your theme:

## Option 1 - Initialize with vanilla JavaScript

Go to [[Publii sites directory](https://getpublii.com/docs/publii-global-settings.html#fileslocation)] > [site folder] > input > [theme or [theme-override](https://getpublii.com/dev/theme-overrides/)] > index.hbs

➡️ add `class` for articles

Edit from this:

```handlebars
      <div class="posts">
         {{#each posts}}
            <article>
```


to this:

```handlebars
      <div class="posts">
         {{#each posts}}
            <article class="c-card">
```


Go to [[Publii sites directory](https://getpublii.com/docs/publii-global-settings.html#fileslocation)] > [site folder] > input > [theme or [theme-override](https://getpublii.com/dev/theme-overrides/)] > partials > pagination.hbs

➡️ add `class` for pagination links

Edit from this (2 occurrences):

```handlebars
      {{#if @pagination.nextPage}}
         <a href="{{@pagination.nextPageUrl}}">
```


to this:

```handlebars
      {{#if @pagination.nextPage}}
         <a class="pagination__next" href="
```


Then, in the Publii app: go to Tools & Plugins > Custom HTML > Footer and paste the following below the infinite-scroll.js:

```html
<script>
let elem = document.querySelector('.posts');
let infScroll = new InfiniteScroll( elem, {
  // options
  path: '.pagination__next',
  append: '.c-card',
  history: false,
});
</script>
```

## Option 2 - Initialize with HTML

(Assuming you added the classes in option 1) Go to [[Publii sites directory](https://getpublii.com/docs/publii-global-settings.html#fileslocation)] > [site folder] > input > [theme or [theme-override](https://getpublii.com/dev/theme-overrides/)] > index.hbs

Edit from this:

```handlebars
<div class="posts">
```

To this:

```handlebars
<div class="posts" data-infinite-scroll='{ "path": ".pagination__next", "append": ".c-card", "history": false }'>
```

**Reply by pvanthony** [reply]
Thank you for the detail steps on how to do it.

**Comment by davidrutland** [comment]
Hey @pvanthony,  I had some success using infinite scroll to append featured articles to the ends of posts. It works pretty well: [https://thecrow.uk/infinite-featured-post-scroll-with-publii-static-site-generator/](https://thecrow.uk/infinite-featured-post-scroll-with-publii-static-site-generator/), and should be pretty adaptable for your use case.

**Reply by pvanthony** [reply]
This looks good. Thank you for sharing the solution too.
